rllib/models/specs/checker.py

```diff
@@ -0,0 +1,290 @@
+
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.models.specs.specs_base import Spec, TypeSpec
+from ray.rllib.models.specs.specs_dict import SpecDict
+from ray.rllib.models.specs.typing import SpecType
+
+
+def _convert_to_canonical_format(spec: SpecType) -> Union[Spec, SpecDict]:
+    """Converts a spec type input to the canonical format.
+
+    The canonical format is either a tree structure with SpecType leaves (i.e.
+    SpecDict) or if the spec is a single constraint, a single SpecType object.
+
+    The input can be any of the following:
+        - a list of nested_keys. nested_keys are either strings or tuples of strings
+        specifying the path to a leaf in the tree.
+        - a tree of constraints. The tree structure can be specified as any nested
+        hash-map structure (e.g. dict, SpecDict, NestedDict, etc.) The leaves of the
+        tree can be either a Spec object, a type, or None. If the leaf is a type, it is
+        converted to a TypeSpec. If the leaf is None, only the existance of the key is
+        checked.
+        - a single constraint. The constraint can be a Spec object, a type, or None.
+
+    Args:
+        spec: The spec to convert to canonical format.
+
+    Returns:
+        The canonical format of the spec.
+    """
+    # convert spec of form list of nested_keys to model_spec with None leaves
+    if isinstance(spec, list):
+        spec = [(k,) if isinstance(k, str) else k for k in spec]
+        return SpecDict({k: None for k in spec})
+
+    # convert spec of form tree of constraints to model_spec
+    if isinstance(spec, abc.Mapping):
+        spec = SpecDict(spec)
+        for key in spec:
+            # if values are types or tuple of types, convert to TypeSpec
+            if isinstance(spec[key], (type, tuple)):
+                spec[key] = TypeSpec(spec[key])
+        return spec
+
+    if isinstance(spec, type):
+        return TypeSpec(spec)
+
+    # otherwise, assume spec is already in canonical format
+    return spec
+
+
+def _should_validate(
+    cls_instance: object, method: Callable, tag: str = "input"
+) -> bool:
+    """Returns True if the spec should be validated, False otherwise.
+
+    The spec should be validated if the method is not cached (i.e. there is no cache
+    storage attribute in the instance) or if the method is already cached. (i.e. it
+    exists in the cache storage attribute)
+
+    Args:
+        cls_instance: The class instance that the method belongs to.
+        method: The method to apply the spec checking to.
+        tag: The tag of the spec to check. Either "input" or "output". This is used
+        internally to defined an internal cache storage attribute based on the tag.
+
+    Returns:
+        True if the spec should be validated, False otherwise.
+    """
+    cache_store = getattr(cls_instance, f"__checked_{tag}_specs_cache__", None)
+    return cache_store is None or method.__name__ not in cache_store
+
+
+def _validate(
```

Comment:
Why do we have underscores instead of (at)DeveloperAPI in front of these?

Reply:
Fine with me as long as we don't need to import that function (that starts with a "_") somewhere else.

Reply:
yes. It should not be imported elsewhere.